src/Relation/Binary/Properties/HeytingAlgebra.agda

```diff
@@ -0,0 +1,101 @@
+open import Relation.Binary.Properties.JoinSemilattice joinSemilattice
+open import Relation.Binary.Properties.Lattice lattice
+
+private
+  ≈-setoid : Setoid _ _
+  ≈-setoid = record { isEquivalence = isEquivalence }
+
+open Setoid ≈-setoid using () renaming (sym to ≈-sym; refl to ≈-refl)
+
+
+⇨-eval : ∀ x y → (x ⇨ y) ∧ x ≤ y
+⇨-eval x y = let _ , transposeʳ = exponential (x ⇨ y) x y
+             in transposeʳ refl
+
+⇨-∧-distribʳ : _⇨_ DistributesOverˡ _∧_
+⇨-∧-distribʳ x y z = antisym
+  (let _     , _     , least = infimum (x ⇨ y) (x ⇨ z)
+       app-y , _             = exponential (x ⇨ y ∧ z) x y
+       app-z , _             = exponential (x ⇨ y ∧ z) x z
+       y∧z≤y , y∧z≤z , _     = infimum y z
+   in least _ (app-y (trans (⇨-eval x (y ∧ z)) y∧z≤y))
+              (app-z (trans (⇨-eval x (y ∧ z)) y∧z≤z)))
+   let app   , _             = exponential ((x ⇨ y) ∧ (x ⇨ z)) x (y ∧ z)
+   in app (begin ((x ⇨ y) ∧ (x ⇨ z)) ∧ x       ≈⟨ ∧-cong ≈-refl (≈-sym (∧-idempotent _)) ⟩
+                 ((x ⇨ y) ∧ (x ⇨ z)) ∧ x ∧ x   ≈⟨ ≈-sym (∧-assoc _ _ _) ⟩
+                 (((x ⇨ y) ∧ (x ⇨ z)) ∧ x) ∧ x ≈⟨ ∧-cong (∧-assoc _ _ _) ≈-refl ⟩
+                 ((x ⇨ y) ∧ (x ⇨ z) ∧ x) ∧ x   ≈⟨ ∧-cong (∧-cong ≈-refl (∧-comm _ _)) ≈-refl ⟩
+                 ((x ⇨ y) ∧ x ∧ (x ⇨ z)) ∧ x   ≈⟨ ∧-cong (≈-sym (∧-assoc _ _ _)) ≈-refl ⟩
+                 (((x ⇨ y) ∧ x) ∧ (x ⇨ z)) ∧ x ≈⟨ ∧-assoc _ _ _ ⟩
+                 ((x ⇨ y) ∧ x) ∧ (x ⇨ z) ∧ x   ≤⟨ ∧-monotonic (⇨-eval _ _) (⇨-eval _ _) ⟩
+                 y ∧ z                         ∎)
+   where open PoR poset
+
+∧-∨-distribˡ : _∧_ DistributesOverˡ _∨_
+∧-∨-distribˡ x y z = antisym
+  (let rhs                      = (x ∧ y ∨ x ∧ z)
+       _     , transposeʳ       = exponential (y ∨ z) x rhs
+       _     , _     , greatest = supremum y z
+       app-y , _                = exponential y x rhs
+       app-z , _                = exponential z x rhs
+   in trans (reflexive (∧-comm _ _))
+            (transposeʳ (greatest _ (app-y (trans (reflexive (∧-comm _ _))
+                                                  (proj₁ (supremum _ _))))
+                                    (app-z (trans (reflexive (∧-comm _ _))
+                                                  (proj₁ (proj₂ (supremum _ _))))))))
+  let _      , _     , least    = infimum x (y ∨ z)
+      _      , _     , greatest = supremum (x ∧ y) (x ∧ z)
+      x∧y≤x  , x∧y≤y , _        = infimum x y
+      x∧z≤x  , x∧z≤z , _        = infimum x z
+      y≤y∨z  , z≤y∨z , _        = supremum y z
+  in least _ (greatest _ x∧y≤x x∧z≤x)
+             (greatest _ (trans x∧y≤y y≤y∨z)
+                         (trans x∧z≤z z≤y∨z))
+
+
+isDistributiveLattice : IsDistributiveLattice _≈_ _≤_ _∨_ _∧_
+isDistributiveLattice = record { isLattice = isLattice ; ∧-∨-distribˡ = ∧-∨-distribˡ }
+
+distributiveLattice : DistributiveLattice _ _ _
+distributiveLattice = record { isDistributiveLattice = isDistributiveLattice }
+
+∨-⇨-distrib : ∀ x y z → x ∨ y ⇨ z ≈ (x ⇨ z) ∧ (y ⇨ z)
```

Comment:
this name is not very good, but I can't come up with a better one. let me know what's your suggestions.

Reply:
We've usually used the naming convention of `⇨-distribˡ-∨-∧` for properties similar to these.